blockprod/src/detail/job_manager.rs

```diff
@@ -0,0 +1,358 @@
+                                ChainstateEvent::NewTip(block_id, _) => {
+                                    if let Err(e) = chainstate_sender.send(block_id.into()) {
+                                        log::error!(
+                                            "Chainstate subscriber failed to send new tip: {e:?}"
+                                        )
+                                    }
+                                }
+                            },
+                        );
+
+                    this.subscribe_to_events(subscribe_func);
+                })
+                .await
+        });
+    }
+
+    fn chainstate_handler(jobs: &mut BTreeMap<JobKey, JobHandle>, event: Option<Id<GenBlock>>) {
+        if let Some(new_tip_id) = event {
+            let mut jobs_to_stop: Vec<JobKey> = vec![];
+
+            for (job_key, _) in jobs.iter() {
+                if job_key.current_tip_id() != new_tip_id {
+                    jobs_to_stop.push(job_key.clone());
+                }
+            }
+
+            for job_key in jobs_to_stop {
+                if let Some(handle) = jobs.remove(&job_key) {
+                    if let Err(e) = handle.cancel_sender.send(()) {
+                        log::info!("Error sendin cancel job event: {e:?}");
+                    }
+                }
+            }
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn get_job_count(&self) -> Result<usize, JobManagerError> {
+        let (result_sender, result_receiver) = oneshot::channel();
+
+        if self.get_job_count_sender.send(result_sender).is_err() {
+            return Err(JobManagerError::FailedToSendGetJobCountEvent);
+        }
+
+        tokio::select! {
+            result = result_receiver => match result {
+                Ok(v) => Ok(v),
+                Err(_) => Err(JobManagerError::FailedToReadJobCount),
+            }
+        }
+    }
+
+    fn get_job_count_handler(
+        jobs: &BTreeMap<JobKey, JobHandle>,
+        event: Option<oneshot::Sender<usize>>,
+    ) {
+        if let Some(result_sender) = event {
+            if let Err(e) = result_sender.send(jobs.len()) {
+                log::info!("Error sending get job count results: {e:?}");
+            }
+        }
+    }
+
+    pub async fn new_job(
+        &self,
+        block_id: Id<GenBlock>,
+    ) -> Result<(JobKey, UnboundedReceiver<()>), JobManagerError> {
+        let (result_sender, result_receiver) = oneshot::channel();
+        let (cancel_sender, cancel_receiver) = unbounded_channel::<()>();
+
+        if self.new_job_sender.send((block_id, cancel_sender, result_sender)).is_err() {
+            return Err(JobManagerError::FailedToSendNewJobEvent);
+        }
+
+        tokio::select! {
+            result = result_receiver => match result {
+                Ok(v) => match v {
+                    Ok(v) => Ok((v, cancel_receiver)),
+                    Err(e) => Err(e),
+                },
```

Comment:
Possibly something like
```suggestion
                Ok(v) => v.map(move |v| (v, cancel_receiver)),
```

Reply:
Nice. This is now done.